src/hotspot/share/gc/shenandoah/shenandoahRootProcessor.inline.hpp

```diff
@@ -81,43 +81,46 @@ ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::ShenandoahClassLoad
   if (!SINGLE_THREADED) {
     ClassLoaderDataGraph::clear_claimed_marks();
   }
-  if (CONCURRENT) {
+  if (CONCURRENT && !SINGLE_THREADED) {
     ClassLoaderDataGraph_lock->lock();
   }
+
+  // Non-concurrent mode only runs at safepoints by VM thread
+  assert(CONCURRENT || SafepointSynchronize::is_at_safepoint(), "Must be at a safepoint");
+  assert(CONCURRENT || Thread::current()->is_VM_thread(), "Can only be done by VM thread");
 }
 
 template <bool CONCURRENT, bool SINGLE_THREADED>
 ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::~ShenandoahClassLoaderDataRoots() {
-  if (CONCURRENT) {
+  if (CONCURRENT && !SINGLE_THREADED) {
     ClassLoaderDataGraph_lock->unlock();
   }
 }
 
+template <bool CONCURRENT, bool SINGLE_THREADED>
+void ShenandoahClassLoaderDataRoots<CONCURRENT, SINGLE_THREADED>::cld_do_impl(CldDo f, CLDClosure* clds, uint worker_id) {
+  if (CONCURRENT) {
+    if (_semaphore.try_acquire()) {
+      ShenandoahWorkerTimingsTracker timer(_phase, ShenandoahPhaseTimings::CLDGRoots, worker_id);
+      if (SINGLE_THREADED) ClassLoaderDataGraph_lock->lock();
+      f(clds);
+      if (SINGLE_THREADED) ClassLoaderDataGraph_lock->unlock();
+      _semaphore.claim_all();
```

Comment:
Would it be nicer to write this as follows?
if (SINGLE_THREADED) {
  MutexLocker ml(ClassLoaderDataGraph_lock);
  f(clds);
} else {
  f(clds);
}